influxdb3_cache/src/meta_cache/cache.rs

```diff
@@ -0,0 +1,513 @@
+                    // repeated `count` times, i.e., depending on how many values come out of
+                    // subsequent nodes:
+                    let block = builder.append_block(value.0.as_bytes().into());
+                    for _ in 0..count {
+                        builder
+                            .try_append_view(block, 0u32, value.0.as_bytes().len() as u32)
+                            .expect("append view for known valid block, offset and length");
+                    }
+                    total_count += count;
+                }
+            } else {
+                builder.append_value(value.0);
+                total_count += 1;
+            }
+        }
+        total_count
+    }
+
+    /// Evaluate a predicate against a [`Node`], producing a list of [`Value`]s and, if this is a
+    /// branch node in the cache tree, a reference to the next [`Node`].
+    fn evaluate_predicate(
+        &self,
+        expired_time_ns: i64,
+        predicate: &Predicate,
+    ) -> Vec<(Value, Option<&Node>)> {
+        match &predicate {
+            Predicate::In(in_list) => in_list
+                .iter()
+                .filter_map(|v| {
+                    self.0.get_key_value(v).and_then(|(v, (t, n))| {
+                        (t > &expired_time_ns).then(|| (v.clone(), n.as_ref()))
+                    })
+                })
+                .collect(),
+            Predicate::NotIn(not_in_set) => self
+                .0
+                .iter()
+                .filter(|(v, (t, _))| t > &expired_time_ns && !not_in_set.contains(v))
+                .map(|(v, (_, n))| (v.clone(), n.as_ref()))
+                .collect(),
+        }
+    }
+}
+
+/// A cache value, which for now, only holds strings
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Hash)]
+pub(crate) struct Value(Arc<str>);
+
+impl From<&FieldData> for Value {
+    fn from(field: &FieldData) -> Self {
+        match field {
+            FieldData::Key(s) => Self(Arc::from(s.as_str())),
+            FieldData::Tag(s) => Self(Arc::from(s.as_str())),
+            FieldData::String(s) => Self(Arc::from(s.as_str())),
+            FieldData::Timestamp(_)
+            | FieldData::Integer(_)
+            | FieldData::UInteger(_)
+            | FieldData::Float(_)
+            | FieldData::Boolean(_) => panic!("unexpected field type for metadata cache"),
+        }
+    }
+}
+
+impl From<String> for Value {
+    fn from(value: String) -> Self {
+        Self(Arc::from(value.as_str()))
+    }
+}
+
+/// A predicate that can be applied when gathering [`RecordBatch`]es from a [`MetaCache`]
+///
+/// This is intended to be derived from a set of filter expressions in Datafusion by analyzing
+/// them with a `LiteralGuarantee`.
+///
+/// This uses a `BTreeSet` to store the values so that they are iterated over in sorted order.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub(crate) enum Predicate {
+    In(BTreeSet<Value>),
+    NotIn(BTreeSet<Value>),
+}
```

Comment:
This type was changed from the previous `enum` that also contained `Eq` and `NotEq` variants. Those were no longer needed with how predicates are pushed down from DataFusion (see the `table_function.rs` module)